lib/router/src/lib.rs

```diff
@@ -208,43 +247,93 @@ where
         use svc::Service;
 
         loop {
             self.state = match self.state {
-                State::Init(ref mut req, ref mut svc) => {
-                    match svc.poll_ready().map_err(error::Error::from)? {
-                        Async::NotReady => {
-                            unreachable!("load shedding services must always be ready");
+                State::Inserting(ref mut rr) => {
+                    // Aquire the lock for the router cache
+                    let mut cache = match rr.cache.poll_lock() {
+                        Async::Ready(aquired) => aquired,
+                        Async::NotReady => return Ok(Async::NotReady),
+                    };
+
+                    let request = rr.request.take().expect("polled after ready");
+                    let target = rr.target.take().expect("polled after ready");
+
+                    // If the target is already cached, route the request to
+                    // the service; otherwise, try to insert it
+                    if let Some(service) = cache.get(&target) {
+                        State::Calling(Some(request), Some(service))
+                    } else {
+                        // Ensure that there is capacity for a new slot
+                        match cache.poll_insert().expect("Cache::reserve must not fail") {
+                            Async::Ready(()) => {
+                                // Make a new service for the target
+                                let make = rr.make.take().expect("polled after ready");
+                                let service = LoadShed::new(make.make(&target));
+
+                                cache.insert(target, service.clone());
+                                State::Calling(Some(request), Some(service))
+                            }
+                            Async::NotReady => {
+                                let capacity = cache.capacity();
+                                State::Error(Some(error::NoCapacity(capacity).into()))
+                            }
+                        }
+                    }
+                }
+                State::Calling(ref mut request, ref mut service) => {
+                    let mut service = service.take().expect("polled after ready");
+
+                    match service.poll_ready() {
+                        Ok(Async::Ready(())) => {
+                            let request = request.take().expect("polled after ready");
+                            State::Called(service.call(request))
                         }
-                        Async::Ready(()) => {
-                            let req = req.take().expect("polled after ready");
-                            State::Called(svc.call(req))
+                        Ok(Async::NotReady) => {
+                            unreachable!("load shedding services must always be ready");
                         }
+                        Err(err) => State::Error(Some(err)),
```

Comment:
Similarly, if we just returned the error directly here, I think we could say `match service.poll_ready()?` above, avoiding this match arm...